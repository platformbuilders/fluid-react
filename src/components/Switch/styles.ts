import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  text-align: center;
`;

export const Input = styled.input`
  display: inline-block;
  position: relative;
  margin: 0 0 10px;
  font-size: 16px;
  line-height: 24px;
  position: absolute;
  top: 0;
  left: 0;
  width: 36px;
  height: 20px;
  opacity: 0;
  z-index: 0;
`;

type LabelProps = {
  $checked: boolean;
  $variant?: 'default' | 'expanded';
};

export const Label = styled.label<LabelProps>`
  display: ${({ $variant }) => ($variant === 'expanded' ? 'flex' : 'block')};
  position: relative;
  margin: 0 0 15px;
  font-size: 16px;
  line-height: 24px;
  padding: 0 0 0 60px;
  cursor: pointer;

  ${({ $variant }) =>
    $variant === 'expanded' &&
    css`
      align-items: center;
      height: 32px;
    `}

  &:before {
    content: '';
    position: absolute;
    top: ${({ $variant }) => ($variant === 'expanded' ? '0' : '5px')};
    left: 0;
    z-index: 1;
    transition: background-color 0.28s cubic-bezier(0.4, 0, 0.2, 1);
    width: ${({ $variant }) => ($variant === 'expanded' ? '52px' : '36px')};
    height: ${({ $variant }) => ($variant === 'expanded' ? '32px' : '14px')};
    background-color: ${({ $checked, $variant }) => {
      if ($checked) {
        return $variant === 'expanded' ? '#1E1F44' : 'rgba(63, 81, 181, 0.5)';
      }
      return 'rgba(0, 0, 0, 0.26)';
    }};
    border-radius: ${({ $variant }) =>
      $variant === 'expanded' ? '999px' : '14px'};
  }

  &:after {
    content: '';
    position: absolute;
    top: 2px;
    left: ${({ $variant, $checked }) => {
      if ($checked) {
        return $variant === 'expanded' ? '22px' : '16px';
      }
      return $variant === 'expanded' ? '2px' : '0';
    }};
    width: ${({ $variant }) => ($variant === 'expanded' ? '28px' : '20px')};
    height: ${({ $variant }) => ($variant === 'expanded' ? '28px' : '20px')};
    background-color: ${({ $variant, $checked }) =>
      $checked && $variant === 'default' ? '#3f51b5' : '#fff'};
    border-radius: 14px;
    box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
      0 3px 1px -2px rgba(0, 0, 0, 0.2), 0 1px 5px 0 rgba(0, 0, 0, 0.12);
    z-index: 2;
    transition: all 0.28s cubic-bezier(0.4, 0, 0.2, 1);
    transition-property: left, background-color;
  }
`;
