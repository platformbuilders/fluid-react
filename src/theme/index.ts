export { default as colors } from './colors';
export { default as radius } from './radius';
export { default as animations } from './animations';
