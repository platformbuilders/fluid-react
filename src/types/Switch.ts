import { ButtonVariants } from '@platformbuilders/theme-toolkit';

export type SwitchProps = {
  checkedIcon?: JSX.Element;
  variant?: ButtonVariants;
  disabled?: boolean;
  icon?: JSX.Element;
  size?: 'small' | 'medium';
  value?: unknown;
  label?: string;
  labelPlacement?: 'top' | 'start' | 'bottom' | 'end';
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
};
