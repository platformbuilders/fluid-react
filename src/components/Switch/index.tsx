import { SwitchProps } from '../../types';
import { StyledSwitch } from './styles';

const Switch = ({ className, ...rest }: SwitchProps) => {
  return <StyledSwitch className={className} {...rest} />;
};

export default Switch;
