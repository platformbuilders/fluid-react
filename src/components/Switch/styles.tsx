import React from 'react';
import styled from 'styled-components';
import { FormControlLabel, Switch } from '@material-ui/core';
import { SwitchProps } from '../../types';
import { getBackgroundColor } from '../../utils/helpers/color';

export const StyledSwitch = styled(({ labelPlacement, label, ...rest }) => (
  <FormControlLabel
    control={<Switch {...rest} />}
    labelPlacement={labelPlacement}
    label={label}
  />
))<SwitchProps>`
  &.MuiSwitch-switchBase.Mui-checked {
    color: ${getBackgroundColor};
    &:hover {
      background-color: ${getBackgroundColor}70;
    }
  }

  &.MuiSwitch-switchBase.Mui-checked + .MuiSwitch-track {
    background-color: ${getBackgroundColor},
  },
`;
