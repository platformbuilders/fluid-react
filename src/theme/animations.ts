import LoadingAnimation from '../assets/animations/loading.json';
import { Animations } from '../types';

export default {
  circularLoading: LoadingAnimation,
  buttonLoading: LoadingAnimation,
  linearLoading: LoadingAnimation,
} as Animations;
