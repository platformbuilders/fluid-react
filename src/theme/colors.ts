import { lighten, darken } from 'polished';
import { Colors } from '../utils/types';

const primary = '#C6B09E';
const secondary = '#805281';
const tertiary = '#2A2E39';
const accent = '#547E6A';

const colors: Colors = {
  failure: '#cc0000',
  warning: '#F5B800',
  success: '#357a38',
  transparent: 'transparent',
  disabled: '#EBEBE4',
  primary: {
    light: lighten(0.05, primary),
    main: primary,
    dark: darken(0.12, primary),
    contrast: '#ffffff',
  },
  secondary: {
    light: lighten(0.05, secondary),
    main: secondary,
    dark: darken(0.1, secondary),
    contrast: '#ffffff',
  },
  tertiary: {
    light: lighten(0.15, tertiary),
    main: tertiary,
    dark: darken(0.1, tertiary),
    contrast: '#ffffff',
  },
  accent: {
    light: lighten(0.05, accent),
    main: accent,
    dark: darken(0.05, accent),
    contrast: '#ffffff',
  },
};

export default colors;
