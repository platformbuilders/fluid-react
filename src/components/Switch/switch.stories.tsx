import { expect, jest } from '@storybook/jest';
import type { Meta, StoryObj } from '@storybook/react';
import { configure, userEvent, within } from '@storybook/testing-library';
import Switch from './index';

configure({ testIdAttribute: 'id' });

//  Mocks
const mockTextId = 'switch-test-id';
const mockOnChange = jest.fn();

const meta: Meta<typeof Switch> = {
  title: 'Components/Switch',
  component: Switch,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    id: mockTextId,
    label: 'Switch Test',
    checked: false,
    onChange: mockOnChange,
    variant: 'default',
  },
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);

    await step('Switch | Test Render', async () => {
      expect(canvas.getByTestId(mockTextId)).toBeInTheDocument();
    });

    await step('Switch | Event Click', async () => {
      await userEvent.click(canvas.getByTestId(mockTextId));
      expect(mockOnChange).toHaveBeenCalled();
    });
  },
};

type Story = StoryObj<typeof Switch>;

export const Default: Story = {
  args: {},
};

export default meta;
