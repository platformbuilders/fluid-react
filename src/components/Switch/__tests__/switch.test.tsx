import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';

import Switch from '..';
import theme from '../../../theme';

describe('Component: Switch', () => {
  test('snapshots', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Switch title="AnÃ´nimo?" value={false} variant="secondary" />
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });
});
