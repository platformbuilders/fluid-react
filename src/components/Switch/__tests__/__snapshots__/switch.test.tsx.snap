// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Switch snapshots 1`] = `
<div>
  <label
    class="MuiFormControlLabel-root"
  >
    <span
      class="MuiSwitch-root sc-bczRLJ bSnHEo"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-1 MuiSwitch-switchBase MuiSwitch-colorSecondary"
        title="AnÃ´nimo?"
        variant="secondary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-4 MuiSwitch-input"
            type="checkbox"
            value="false"
          />
          <span
            class="MuiSwitch-thumb"
          />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    />
  </label>
</div>
`;
